[WebGPU] A retained MTLSamplerState can result us in exceeding the MTLSamplerState limit

mwyrzykowski opened the pull request:
#### 1eb32c1b3c5eaf46ef64a1852e73c845bd755f5f
<pre>
[WebGPU] A retained MTLSamplerState can result us in exceeding the MTLSamplerState limit
<a href="https://bugs.webkit.org/show_bug.cgi?id=279542">https://bugs.webkit.org/show_bug.cgi?id=279542</a>
radar://135708279

Reviewed by Dan Glastonbury.

We were tracking internally the number of samplers to ensure we didn&apos;t exceed
the process limit for MTLSamplerState instances.

This is problematic when a MTLSamplerState is retained outside the cache, e.g.,
in a pending MTLCommandBuffer.

Handle this by overriding dealloc in order to track the real alive count.

* LayoutTests/fast/webgpu/nocrash/fuzz-279542-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-279542.html: Added.
* Source/WebGPU/WebGPU/Sampler.h:
* Source/WebGPU/WebGPU/Sampler.mm:
(WebGPU::Sampler::deallocSamplerState):
(WebGPU::Sampler::Sampler):
(WebGPU::Sampler::samplerState const):

Canonical link: <a href="https://commits.webkit.org/283645@main">https://commits.webkit.org/283645@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/504210eb085f69892bf74180da9a5c7810606731

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46176 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70836 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17934 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53975 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17710 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53511 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12070 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69868 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42485 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57791 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34145 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39157 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15183 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16288 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61069 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15524 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72537 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14881 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60920 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10790 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57848 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61154 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14779 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8832 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2452 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41983 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43060 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44243 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42803 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg reviewed (approved):
:shipit: with a 🤨 at adding a new selector to `_MTLSamplerState`

mwyrzykowski committed:
[WebGPU] A retained MTLSamplerState can result us in exceeding the MTLSamplerState limit
https://bugs.webkit.org/show_bug.cgi?id=279542
radar://135708279

Reviewed by Dan Glastonbury.

We were tracking internally the number of samplers to ensure we didn't exceed
the process limit for MTLSamplerState instances.

This is problematic when a MTLSamplerState is retained outside the cache, e.g.,
in a pending MTLCommandBuffer.

Handle this by overriding dealloc in order to track the real alive count.

* LayoutTests/fast/webgpu/nocrash/fuzz-279542-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-279542.html: Added.
* Source/WebGPU/WebGPU/Sampler.h:
* Source/WebGPU/WebGPU/Sampler.mm:
(WebGPU::Sampler::deallocSamplerState):
(WebGPU::Sampler::Sampler):
(WebGPU::Sampler::samplerState const):

Canonical link: https://commits.webkit.org/283645@main
